src/core/lib/channel/metrics.h

```diff
@@ -0,0 +1,217 @@
+#define GRPC_SRC_CORE_LIB_CHANNEL_METRICS_H
+
+#include <grpc/support/port_platform.h>
+
+#include <atomic>
+#include <cstdint>
+#include <memory>
+#include <vector>
+
+#include "absl/strings/string_view.h"
+
+#include <grpc/support/log.h>
+
+#include "src/core/lib/channel/channel_args.h"
+#include "src/core/lib/gprpp/sync.h"
+
+namespace grpc_core {
+
+// Static only.
+class GlobalInstrumentsRegistry {
+ public:
+  enum class ValueType {
+    kUndefined,
+    kUInt64,
+    kDouble,
+  };
+  enum class InstrumentType {
+    kUndefined,
+    kCounter,
+    kHistogram,
+  };
+  struct GlobalInstrumentDescriptor {
+    ValueType value_type;
+    InstrumentType instrument_type;
+    uint32_t index;
+    absl::string_view name;
+    absl::string_view description;
+    absl::string_view unit;
+    std::vector<absl::string_view> label_keys;
+    std::vector<absl::string_view> optional_label_keys;
+  };
+  struct GlobalHandle {
+    // This is the index for the corresponding registered instrument that
+    // StatsPlugins can use to uniquely identify an instrument in the current
+    // process. Though this is not guaranteed to be stable between different
+    // runs or between different versions.
+    uint32_t index;
+  };
+  struct GlobalUInt64CounterHandle : GlobalHandle {};
+  struct GlobalDoubleCounterHandle : GlobalHandle {};
+  struct GlobalUInt64HistogramHandle : GlobalHandle {};
+  struct GlobalDoubleHistogramHandle : GlobalHandle {};
+
+  // Creates instrument in the GlobalInstrumentsRegistry.
+  static GlobalUInt64CounterHandle RegisterUInt64Counter(
+      absl::string_view name, absl::string_view description,
+      absl::string_view unit, std::vector<absl::string_view> label_keys,
+      std::vector<absl::string_view> optional_label_keys);
+  static GlobalDoubleCounterHandle RegisterDoubleCounter(
+      absl::string_view name, absl::string_view description,
+      absl::string_view unit, std::vector<absl::string_view> label_keys,
+      std::vector<absl::string_view> optional_label_keys);
+  static GlobalUInt64HistogramHandle RegisterUInt64Histogram(
+      absl::string_view name, absl::string_view description,
+      absl::string_view unit, std::vector<absl::string_view> label_keys,
+      std::vector<absl::string_view> optional_label_keys);
+  static GlobalDoubleHistogramHandle RegisterDoubleHistogram(
+      absl::string_view name, absl::string_view description,
+      absl::string_view unit, std::vector<absl::string_view> label_keys,
+      std::vector<absl::string_view> optional_label_keys);
+  // Getter functions for StatsPlugins to query registered counters/histograms.
+  static const std::vector<GlobalInstrumentDescriptor>& instruments();
+
+  GlobalInstrumentsRegistry() = delete;
+
+ private:
+  static std::vector<GlobalInstrumentDescriptor> instruments_;
+};
+
+// Interface.
+class StatsPlugin {
+ public:
+  virtual ~StatsPlugin() = default;
+  virtual bool IsEnabledForTarget(absl::string_view target) = 0;
```

Comment:
All of this wrapping seems a little silly, since we really want the same type in both layers.  Why not just have a single type that we use in both places?

Reply:
We can add a `ChannelScope` in the public interface and it can be used by both the `StatsPlugin` interface and `OpenTelemetryPluginBuilder`. Or we can just add it in `StatsPlugin` if we want to just make `StatsPlugin` public. @yashykt

Reply:
I think you'll need to define it in a C-core header in the `grpc_core` namespace.  Then, in the C++ header, include the C-core header and make an alias for the type in the `grpc` namespace via a `using` statement.

Reply:
@yashykt suggested that if we do this we might accidentally introduce API breaking changes while the wrapping method would allow us to catch it.

Reply:
Yeah, I'm a bit worried as to what happens when we make any changes to the Core `Scope` class, but maybe we can avoid this issue for now, by adding comments saying that this class should not make any breaking changes since it's being by C++ public API.

I still think wrapping it would be safer as we do for all C++ types that use Core types underneath.

Reply:
In the long run, there isn't really going to be much difference between the C-core API and the C++ API, and we're trying to minimize this kind of wrapping whenever possible.  We've actually already started using this non-wrapping pattern for a few things (e.g., see #32750 and #34715).

I think we're going to need to get used to the fact that we can't change C-core APIs that affect the public API.  I don't think we should do something different for this particular API than we're going to do for any other.

Reply:
Got it, thanks for the explanation. I have created `class ChannelScope` under the `grpc_core` namespace. I will make an alias for it in the `grpc` namespace later when I add to the public header.